### PR TITLE
[Perf] leverage ovs_control function to improve performance

### PR DESCRIPTION
**Pull request by er1cthe0ne**
To further ehancement performance specially on massive number of neighbor configuration handling, this change leverage ovs_control functions instead of making system call for ovs-ofctl. The result is ~3x improvement for 1000 and 10000 neighbors handling.

The change includes:

1. switched to ovs_control functions for neighbor configuration handling
1. improved ovs_control functions on logging and elapse time recording, added the corresponding test cases
1. built ovs library as ndebug (no debug) to unblock the benign assert
1. tried out release build and fix compile time error for release build, observe minimal performance improvement
1. updated test code to show bridge delete/setup initialization times
1. added log file to record the current performance numbers

### CODE REVIEW

File: docs/logs/perf_N_neighbors_physical_2020_12_31.log [hunk 1]
```diff
@@ -0,0 +1,186 @@
+./build/tests/aca_tests --gtest_filter=*1_port_CREATE_plus_10_l2_neighbor_CREATE* -n [number of L2 neighbors to create]
+build = release
```
Comment (cj-chung) [inline docs/logs/perf_N_neighbors_physical_2020_12_31.log:2]: Do we need to put this perf log file on github?

File: src/ovs/aca_vlan_manager.cpp [hunk 2]
```diff
@@ -174,45 +174,52 @@ int ACA_Vlan_Manager::delete_ovs_port(string /*vpc_id*/, string ovs_port,
...
+                         "->tun_dst,output:" + VXLAN_GENERIC_OUTPORT_NUMBER;
 
-  cmd_string = "add-flow br-tun \"table=51,priority=50,arp,dl_vlan=" +
-               to_string(internal_vlan_id) + ",nw_dst=" + virtual_ip +
-               " actions=move:NXM_OF_ETH_SRC[]->NXM_OF_ETH_DST[],mod_dl_src:" + virtual_mac +
-               ",load:0x2->NXM_OF_ARP_OP[],move:NXM_NX_ARP_SHA[]->NXM_NX_ARP_THA[],move:NXM_OF_ARP_SPA[]->NXM_OF_ARP_TPA[],load:0x" +
-               current_virtual_mac + "->NXM_NX_ARP_SHA[],load:" + string(hex_ip_buffer) +
-               "->NXM_OF_ARP_SPA[],in_port\"";
+  overall_rc = ACA_OVS_Control::get_instance().add_flow(
+          "br-tun", (match_string + action_string).c_str());
```
Comment (cj-chung) [inline src/ovs/aca_vlan_manager.cpp:197]: Maybe we can refactor `add_flow()` method to accept `match_string` part and `action_string` part in the future.

File: src/ovs/ovs_control.cpp [hunk 3]
```diff
@@ -100,654 +100,615 @@ bool OVS_Control::bundle;
...
+    run(vconn_dump_flows(vconn, &fsr, protocol, &fses, &n_fses), "dump flows");
+
+    struct ds s = DS_EMPTY_INITIALIZER;
+    for (size_t i = 0; i < n_fses; i++) {
+      ds_clear(&s);
+      ofputil_flow_stats_format(&s, &fses[i], ports_to_show(bridge), tables_to_show(bridge),
+                                //ports_to_show(ctx->argv[1]),
+                                //tables_to_show(ctx->argv[1]),
+                                show_stats);
+      printf(" %s\n", ds_cstr(&s));
```
Comment (cj-chung) [inline src/ovs/ovs_control.cpp:274]: Should we change `printf` to `ACA_LOG_DEBUG` here?

File: src/ovs/ovs_control.cpp [hunk 4]
```diff
@@ -100,654 +100,615 @@ bool OVS_Control::bundle;
...
+      ;
```
Comment (cj-chung) [inline src/ovs/ovs_control.cpp:275]: Extra `;`.

File: docs/logs/perf_N_neighbors_physical_2020_12_31.log (hunk 1 above)
Comment (er1cthe0ne) [reply docs/logs/perf_N_neighbors_physical_2020_12_31.log:2]: it doesn't hurt for record keeping :)

File: src/ovs/aca_vlan_manager.cpp (hunk 2 above)
Comment (er1cthe0ne) [reply src/ovs/aca_vlan_manager.cpp:197]: yes, if we decide to invest more on ovs datapath.

File: src/ovs/ovs_control.cpp [hunk 5]
```diff
@@ -100,654 +100,615 @@ bool OVS_Control::bundle;
...
+
+  error = parse_ofp_flow_mod_str(&fm, flow, ports_to_accept(bridge),
+                                 tables_to_accept(bridge), command, &usable_protocols);
+  if (error) {
+    // ovs_fatal(0, "%s", error);
+    ACA_LOG_ERROR("%s", error);
+    rc = EXIT_FAILURE;
+  } else {
+    // flow_mod__ returns void
     flow_mod__(bridge, &fm, 1, usable_protocols);
```
Comment (er1cthe0ne) [inline src/ovs/ovs_control.cpp:411]: @cj-chung - we are passing in 1 as the number of flows to flow_mod__ here, therefore, I think it wil require some work to fully light up the "bundle" option.

File: src/ovs/ovs_control.cpp (hunk 3 above)
Comment (er1cthe0ne) [reply src/ovs/ovs_control.cpp:274]: fixed.

File: src/ovs/ovs_control.cpp (hunk 4 above)
Comment (er1cthe0ne) [reply src/ovs/ovs_control.cpp:275]: fixed.

File: src/ovs/ovs_control.cpp (hunk 5 above)
Comment (er1cthe0ne) [reply src/ovs/ovs_control.cpp:411]: for reference, in ovs code, there is parse_ofp_flow_mod_file which has ability to parse multiple flow messages, but I don't like the approach of writing flow message to a file and then read it for bundling.

Comment (cj-chung) [reply src/ovs/ovs_control.cpp:411]: @er1cthe0ne  If flows as bundle come from a file, we need to add the following code or function:
```
static void
ofctl_flow_mod_file(int argc OVS_UNUSED, char *argv[], int command)
{
    enum ofputil_protocol usable_protocols;
    struct ofputil_flow_mod *fms = NULL;
    size_t n_fms = 0;
    char *error;

    if (command == OFPFC_ADD) {
        /* Allow the file to specify a mix of commands.  If none specified at
         * the beginning of any given line, then the default is OFPFC_ADD, so
         * this is backwards compatible. */
        command = -2;
    }
    error = parse_ofp_flow_mod_file(argv[2], ports_to_accept(argv[1]),
                                    tables_to_accept(argv[1]), command,
                                    &fms, &n_fms, &usable_protocols);
    if (error) {
        ovs_fatal(0, "%s", error);
    }
    ofctl_flow_mod__(argv[1], fms, n_fms, usable_protocols);
    free(fms);
}
```
Or we need to count `n_fms` (the size of fms) from the bundle flows.

Comment (cj-chung) [reply src/ovs/ovs_control.cpp:410]: It's better create another PR for the bundle feature improvement.

**Review by cj-chung — APPROVED** [review]
LGTM

**Pull request merged**